### PR TITLE
Implements Package Replier in Twilio

### PR DESCRIPTION
**Pull request by JeffNeff**
Previously this reported a response event type of "io.triggermesh.targets.response" . In reality this target never responded with anything. With this update we are now responding like so :

```
< HTTP/1.1 200 OK
< Ce-Category: success
< Ce-Id: 9bb5902b-19cf-4a1f-9451-b33fb52ec26c
< Ce-Source: twilio-target-adapter
< Ce-Specversion: 1.0
< Ce-Time: 2021-10-01T17:51:25.899401Z
< Ce-Type: io.triggermesh.twilio.sms.send.response
< Content-Length: 43
< Content-Type: application/json
< Date: Fri, 01 Oct 2021 17:51:25 GMT
"message sent to: +‪‪19194465428‬‬"*
```

### CODE REVIEW

File: pkg/targets/adapter/twiliotarget/config.go [hunk 1]
```diff
@@ -27,8 +27,16 @@ func EnvAccessorCtor() pkgadapter.EnvConfigAccessor {
...
-	Token      string `envconfig:"TWILIO_TOKEN" required:"true"`
-	PhoneFrom  string `envconfig:"TWILIO_DEFAULT_FROM" required:"false"`
-	PhoneTo    string `envconfig:"TWILIO_DEFAULT_TO" required:"false"`
+	// Token is the API key for the Twilio account.
+	Token string `envconfig:"TWILIO_TOKEN" required:"true"`
+	// PhoneFrom is the phone number to use as the sender of the SMS
+	PhoneFrom string `envconfig:"TWILIO_DEFAULT_FROM" required:"false"`
+	// PhoneTo is the phone number to send the message to
+	PhoneTo string `envconfig:"TWILIO_DEFAULT_TO" required:"false"`
+	// CloudEvents responses parametrization
```
Comment (odacremolbap) [inline pkg/targets/adapter/twiliotarget/config.go:38]: nit: a blank line between Twilio envs and CloudEvents/Bridge configuration might help reading this block
```suggestion

	// CloudEvents responses parametrization
```

File: pkg/targets/reconciler/twiliotarget/target_adapter.go [hunk 2]
```diff
@@ -83,17 +91,24 @@ func makeAppEnv(spec *v1alpha1.TwilioTargetSpec) []corev1.EnvVar {
...
 		})
 	}
 
+	if spec.EventOptions != nil && spec.EventOptions.PayloadPolicy != nil {
+		env = append(env, corev1.EnvVar{
+			Name:  envEventsPayloadPolicy,
+			Value: string(*spec.EventOptions.PayloadPolicy),
+		})
+	}
+
```
Comment (odacremolbap) [inline pkg/targets/reconciler/twiliotarget/target_adapter.go:112]: Can we add BridgeID propagation to the adapter, like you did here?:

https://github.com/triggermesh/triggermesh/blob/6c43c8a267421e36d33de98d0fcf84744a65347b/pkg/targets/reconciler/alibabaosstarget/adapter.go#L89-L92

**Review by odacremolbap — CHANGES_REQUESTED** [review]
nice, thanks so mucho.

Could you add Bridge ID propagation?

This was probably the first Target ever, and the response "message sent to: +12345" is naive. We could replace that with a structured reply but that can be done in a different PR. Something like this:

```json
{"id":"<incoming-cloudevent-id>", "to": "<sms.To>"}
```

File: pkg/targets/reconciler/twiliotarget/target_adapter.go (hunk 2 above)
Comment (JeffNeff) [reply pkg/targets/reconciler/twiliotarget/target_adapter.go:115]: done in https://github.com/triggermesh/triggermesh/pull/78/commits/d7ca4e644923fad01bc0c396c60b330a091df3a3

**Comment by JeffNeff** [comment]
I thought about changing that.. but I thought that was a little outside the scope of this PR.

**Review by odacremolbap — APPROVED** [review]

File: pkg/targets/adapter/twiliotarget/adapter.go [hunk 3]
```diff
@@ -85,9 +97,10 @@ func (a *twilioAdapter) dispatch(event cloudevents.Event) cloudevents.Result {
 	}
 
 	if _, err := a.client.Messages.SendMessage(sms.From, sms.To, sms.Message, sms.MediaURLs); err != nil {
-		return fmt.Errorf("error sending message: %w", err)
+		return a.replier.Error(&event, targetce.ErrorCodeAdapterProcess, err, nil)
 	}
 
-	a.logger.Infof("Sent message: %s", event.String())
-	return cloudevents.ResultACK
+	info := fmt.Sprintf("message sent to: %s", sms.To)
```
Comment (antoineco) [inline pkg/targets/adapter/twiliotarget/adapter.go:103]: nit: could be slightly simplified as `fmt.Sprint("message sent to ", sms.To)`

File: pkg/targets/adapter/twiliotarget/adapter.go [hunk 4]
```diff
@@ -85,9 +97,10 @@ func (a *twilioAdapter) dispatch(event cloudevents.Event) cloudevents.Result {
...
+	a.logger.Info(info)
```
Comment (antoineco) [inline pkg/targets/adapter/twiliotarget/adapter.go:104]: `Debug(...)` sounds more appropriate here. We should log only errors by default, successes only for troubleshooting.

File: pkg/targets/reconciler/twiliotarget/target_adapter.go [hunk 5]
```diff
@@ -28,7 +28,15 @@ import (
 	servingv1 "knative.dev/serving/pkg/apis/serving/v1"
 )
 
-const adapterName = "twiliotarget"
+const (
+	adapterName = "twiliotarget"
+
+	envTwilioSID           = "TWILIO_SID"
```
Comment (antoineco) [inline pkg/targets/reconciler/twiliotarget/target_adapter.go:34]: Thanks for declaring these constants 👍

**Review by antoineco — COMMENTED** [review]
A minor frown on the debug log at `Info` level :)

File: config/301-twiliotarget.yaml [hunk 6]
```diff
@@ -83,6 +83,13 @@ spec:
                         type: string
                       name:
                         type: string
+              eventOptions:
+                type: object
+                description: "When should this target generate a response event for processing: always, on error, or never"
```
[the remark this answers is not in the document]
Comment (JeffNeff) [reply config/301-twiliotarget.yaml:88]: Because of the `:` we require the quotes

File: pkg/targets/adapter/twiliotarget/adapter.go (hunk 3 above)
Comment (JeffNeff) [reply pkg/targets/adapter/twiliotarget/adapter.go:103]: and then write it twice?

File: pkg/targets/adapter/twiliotarget/adapter.go (hunk 4 above)
Comment (JeffNeff) [reply pkg/targets/adapter/twiliotarget/adapter.go]: fixed!

**Pull request merged**